### ISSUE TITLE
Step result ignored on non-optional question step

### ISSUE THREAD

**Issue by brenbob** [ResearchKit/ResearchKit#294]
I'm initializing a Choice Question step with a previously-saved step result. This works fine, and the question renders with the appropriate choice selected. 

But when I set the question step as non-optional, the step result is ignored. 

My apologies for lack of code example. I'm working to set up a generic example that repro's the problem. Also can check if this occurs for other question types.

**Comment by YuanZhu-apple** [comment]
Can you make a fork and modify ORKText to reproduce this issue?

**Comment by brenbob** [comment]
Not sure about doing a fork, but I can repro the issue with this basic survey implementation. If I set questionStep.optional to 'true', the correct choice is selected on render. If I set it to 'false', no choice is selected.

// main viewController

class SurveyVC: UIViewController, ORKTaskViewControllerDelegate {
    
    /**
    When a task is completed, the `SurveyViewController` calls this closure
    with the created task.
    */
    var taskResultFinishedCompletionHandler: (ORKResult -> Void)?
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        self.startSurvey()
    }
    
    /**
    Start the survey as a RK task
    */
    func startSurvey() {
        
        let task = surveyTask
        
        let taskViewController = SurveyTaskVC(task: task, taskRunUUID: nil)
        
        // Make sure we receive events from `taskViewController`.
        taskViewController.delegate = self
        
        taskViewController.defaultResultSource = taskViewController

        // Assign a directory to store `taskViewController` output.
        taskViewController.outputDirectory = NSURL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(.DocumentDirectory, .UserDomainMask, true)[0] as! String, isDirectory: true)

       // let vc = self.view.window!.rootViewController
 
       self.presentViewController(taskViewController, animated: true, completion: nil)
        
    }
    
 
    var surveyTask: ORKOrderedTask {
        
        var steps = [AnyObject]()

        // Add a question step.
        var choices = [ORKTextChoice]()
        choices.append(ORKTextChoice(text: "YES", value: "1"))
        choices.append(ORKTextChoice(text: "NO", value: "0"))
        var format = ORKTextChoiceAnswerFormat(style: ORKChoiceAnswerStyle.SingleChoice, textChoices: choices)

        let questionStepTitle = "Are you happy?"
        let questionStep = ORKQuestionStep(identifier: "Q1", title: questionStepTitle, answer: format)
        questionStep.optional = false
        
        steps += [questionStep]

        return ORKOrderedTask(identifier: "testTask", steps: steps)
    }
    

    // MARK: ORKTaskViewControllerDelegate
    
    func taskViewController(taskViewController: ORKTaskViewController, didFinishWithReason reason: ORKTaskViewControllerFinishReason, error: NSError?) {
        
        println("didFinishWithReason")
        switch reason {
            
        case ORKTaskViewControllerFinishReason.Completed:
            println("completed")
        case ORKTaskViewControllerFinishReason.Failed:
            println("failed")
        case ORKTaskViewControllerFinishReason.Discarded:
            println("discarded")
        case ORKTaskViewControllerFinishReason.Saved:
            println("saved")
        }
        
        /*
        The `reason` passed to this method indicates why the task view
        controller finished: Did the user cancel, save, or actually complete
        the task; or was there an error?
        
        The actual result of the task is on the `result` property of the task
        view controller.
        */
        taskResultFinishedCompletionHandler?(taskViewController.result)
        
        self.dismissViewControllerAnimated(false, completion: {
            self.performSegueWithIdentifier("backToHome", sender: self)
        } )
        
    }
    
}

import UIKit
import ResearchKit

// taskViewController
class SurveyTaskVC: ORKTaskViewController, ORKTaskResultSource {
    
    func stepResultForStepIdentifier(stepIdentifier: String) -> ORKStepResult? {
        var stepResults = [ORKQuestionResult]()
        if stepIdentifier == "Q1" {
            var questionDefault = ORKChoiceQuestionResult(identifier: stepIdentifier)
            questionDefault.choiceAnswers = ["1"]
            stepResults.append(questionDefault)
        }
        
        var defaults = ORKStepResult(stepIdentifier: stepIdentifier, results: stepResults)
        
        return defaults
    }

    
}

**Comment by YuanZhu-apple** [comment]
@brenbob I would like to debug this code if you can put this with its XCode project into a repo.

**Comment by brenbob** [comment]
Done. https://github.com/brenbob/RKtest

**Comment by YuanZhu-apple** [comment]
Sorry, Xcode project file is missing from your repo.

**Comment by brenbob** [comment]
I've updated the repo to include my Xcode project file

**Comment by YuanZhu-apple** [comment]
Hey @brenbob, I made a few fix to your project run.

Step result from `defaultResultSource` is ignored if you provide a custom stepVC.
Because a result requires to be supplied during a stepVC's init:
```
[[stepViewControllerClass alloc] initWithStep:step result:result];
```

Normally, there is no need to create a custom stepVC for question steps.
But you can hookup the result by yourself in:
```
taskViewController(taskViewController: ORKTaskViewController, viewControllerForStep step: ORKStep) {
   ...
}
```

**Comment by brenbob** [comment]
The custom step is unrelated to this issue and I've removed it from the repo to simplify trouble-shooting.

**Comment by YuanZhu-apple** [comment]
Thanks, I got your point, when a user is visiting a non-optional single choice step already has an answer on it, he/she can't see the answer check mark.

![screen shot 2015-06-22 at 10 08 28 pm](https://cloud.githubusercontent.com/assets/11466704/8299319/a7a60a36-192b-11e5-9006-9f041feb7c72.png)

This behavior is expected (by design). 
Are you suggesting we should change this design? 
What if we make the UI more adaptive for this kind of step (`non-optional single choice step`):
  - Show choice cells with nextButton on step has an answer  (see screen shot bellow)
  - Show arrowed choice cells without next button on step has no answer  (see screen shot above)

![screen shot 2015-06-22 at 10 09 57 pm](https://cloud.githubusercontent.com/assets/11466704/8299440/4b730140-192d-11e5-9ae6-bc4b3803dde3.png)
 
Any other suggestions?

**Comment by brenbob** [comment]
That may work. I'm most concerned about the scenario where the question already has a default answer and the user hits the Back or Cancel button. The saved result is lost and presents a problem in surveys with dynamic task ordering.

**Comment by YuanZhu-apple** [comment]
@brenbob Can you comment on #296 for the saved result lost issue?

**Comment by YuanZhu-apple** [comment]
Fixed #327
